### PR TITLE
[Feature/player 1483] Keyboard navigation for VR 360

### PR DESCRIPTION
**Pull request by IvanSakharovskii**
Navigation into VR video using keyboard

### CODE REVIEW

File: js/components/accessibilityControls.js [hunk 1]
```diff
@@ -3,33 +3,45 @@ var Utils = require('./utils');
 
 var AccessibilityControls = function (controller) {
   this.controller = controller;
+  this.allowed = true;
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Can we use a more descriptive name for this var? From the name it's not clear what exactly is allowed or not.

File: js/components/accessibilityControls.js [hunk 2]
```diff
@@ -61,6 +73,41 @@ AccessibilityControls.prototype = {
...
   },
 
+  keyEventUp: function(e) {
+    if (!this.controller.state.accessibilityControlsEnabled) { return; }
+    var targetTagName = this.getTargetTagName(e);
+    var charCode = e.which || e.keyCode;
+    this.handleKeyClick(e, this.state.keyMapsList, charCode, false, targetTagName);  //stop rotate 360
+  },
+
+  getTargetTagName: function(e) {
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Can we add jsdocs for the new methods?

File: js/components/accessibilityControls.js [hunk 3]
```diff
@@ -3,33 +3,45 @@ var Utils = require('./utils');
...
+  this.allowed = true;
   this.state = {
     seekRate: 1,
-    lastKeyDownTime: 0
+    lastKeyDownTime: 0,
+    "keyMapsList": [
+      {"char": CONSTANTS.KEYCODES.A, "direction": "left"},
+      {"char": CONSTANTS.KEYCODES.D, "direction": "right"},
+      {"char": CONSTANTS.KEYCODES.W, "direction": "up"},
+      {"char": CONSTANTS.KEYCODES.S, "direction": "down"}
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Would this be simpler as a hashmap or do we expect to add more properties apart from `direction`?

```
var keyMapsList = {};
keyMapsList[CONSTANTS.KEYCODES.A] = 'left';
keyMapsList[CONSTANTS.KEYCODES.D] = 'right';
```

File: js/components/accessibilityControls.js [hunk 4]
```diff
@@ -3,33 +3,45 @@ var Utils = require('./utils');
 
 var AccessibilityControls = function (controller) {
   this.controller = controller;
+  this.allowed = true;
   this.state = {
     seekRate: 1,
-    lastKeyDownTime: 0
+    lastKeyDownTime: 0,
+    "keyMapsList": [
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: We should try to keep style consistent. The rest of the file uses single quotes and also doesn't quote property names.

File: js/components/accessibilityControls.js [hunk 5]
```diff
@@ -61,6 +73,41 @@ AccessibilityControls.prototype = {
...
+
+  getTargetTagName: function(e) {
+    var targetTagName = "";
+    if (e.target && typeof e.target.tagName === "string") {
+      targetTagName = e.target.tagName.toLowerCase();
+    }
+    return targetTagName;
+  },
+
+  handleKeyClick: function(e, keyMapsList, char, bool, targetTagName) {
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Can we rename this to something more descriptive? It looks like this is only used for VR, so it might be better to name it according to its purpose. The `e`, `char` and `bool` parameters are also not very descriptive.

File: js/components/accessibilityControls.js [hunk 6]
```diff
@@ -61,6 +73,41 @@ AccessibilityControls.prototype = {
...
+    /*
+     * keyMapsList - array of objects {char: 83, direction: 'down'}
+     */
+    if (!this.controller.videoVr) { return; }
+    for (var i=0; i<keyMapsList.length; i++) {
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: It looks like we can avoid this loop by using the hashmap suggestion above.

File: js/controller.js [hunk 7]
```diff
@@ -643,11 +674,7 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
     },
 
     onBuffering: function(event) {
-      if (this.state.isInitialPlay == false && this.state.screenToShow == CONSTANTS.SCREEN.START_SCREEN) {
-        this.state.buffering = false;
-      } else {
-        this.state.buffering = true;
-      }
+      this.state.buffering = !(this.state.isInitialPlay == false && this.state.screenToShow == CONSTANTS.SCREEN.START_SCREEN);
```
Comment (jpfranco) [inline js/controller.js]: I'm not sure why this is not showing a conflict. This code became out of date after https://github.com/ooyala/html5-skin/pull/825. The value of `this.state.buffering` is no longer set in this function. Please make sure to update your branch with the latest code from master

Comment (jpfranco) [reply js/controller.js]: Actually never mind, I see the conflict now. We generally favor readability, so I recommend just leaving this code as is on the latest version from master.

File: js/components/accessibilityControls.js [hunk 8]
```diff
@@ -61,6 +73,41 @@ AccessibilityControls.prototype = {
...
+      if (char === keyMapsList[i]['char'] && targetTagName !== "button") {
+        if (e.repeat != undefined) {
+          this.allowed = !e.repeat;
+        }
+        if (!this.allowed) {
+          return;
+        }
+        this.allowed = !bool; //prevent repeat of keyDown
+        this.controller.moveToDirection(bool, keyMapsList[i]['direction']);
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Do we need `keyMapsList` as a parameter considering that we're reading other values from `this`?

File: js/components/directionControl.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      var _this = this;
+      this.timer = setTimeout(function(){
+        _this.setState({isTouched: false});
+      }, 300);
+    });
+  },
+  
+  render: function () {
+    var directionClass = 'oo-vr-icon--move--' + this.props.dir
+      , touchedDirectionClass = this.state.isTouched && 'oo-vr-icon--move--' + this.props.dir + '--touched';
```
Comment (jpfranco) [inline js/components/directionControl.js]: Since this a long statement I think it would be more readable as two separate declarations. It would also make it easier to edit without breaking it with a typo.

File: js/components/directionControl.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    return (
+      <div
+        className={classnames('oo-vr-icon--move direction-control', directionClass, touchedDirectionClass)}
+        key={this.props.dir}
+        tabIndex="0"
```
Comment (jpfranco) [inline js/components/directionControl.js]: Is this meant to be accessible via keyboard? If so, then this should  probably be a `button`

File: js/views/playingScreen.js [hunk 11]
```diff
@@ -12,16 +12,22 @@ var React = require('react'),
...
     this.browserSupportsTouch = this.props.controller.state.browserSupportsTouch;
+
     return {
       controlBarVisible: true,
-      timer: null
+      timer: null,
+      isMouseDown: false,
+      isMouseMove: false,
+      XMouseStart: 0,
+      YMouseStart: 0,
```
Comment (jpfranco) [inline js/views/playingScreen.js]: These should be `xMouseStart` and `yMouseStart` for consistent styling.

File: js/views/playingScreen.js [hunk 12]
```diff
@@ -115,16 +109,94 @@ var PlayingScreen = React.createClass({
...
+  },
+  
+  getDirectionParams: function(pageX, pageY) {
+    var dx = pageX - this.state.XMouseStart;
+    var dy = pageY - this.state.YMouseStart;
+    var maxDegreesX = 90;
+    var maxDegreesY = 120;
+    var degreesForPixelYaw = maxDegreesX / this.props.componentWidth;
+    var degreesForPixelPitch = maxDegreesY / this.props.componentHeight;
+    var yaw = (this.props.controller.state.viewingDirection.yaw || 0) + dx * degreesForPixelYaw;
```
Comment (jpfranco) [inline js/views/playingScreen.js]: Do we need validations here, or are these always assumed to be numbers? Would recommend propTypes at least

File: scss/components/_vr-controls.scss [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  @include vendor-prefixes (transition, all .3s ease-in-out);
+  &:hover,
+  &:active,
+  &:focus {
+    .oo-vr-icon--type {
+      -webkit-filter: drop-shadow(-1px 1px 4px rgba(255,255,255,.7));
+      filter: drop-shadow(-1px 1px 4px rgba(255,255,255,.7));
+    }
+  }
+}
```
Comment (jpfranco) [inline scss/components/_vr-controls.scss]: Do we need responsive styles for smaller player sizes?

File: js/components/controlBar.js [hunk 14]
```diff
@@ -22,6 +22,10 @@ var ControlBar = React.createClass({
     this.isMobile = this.props.controller.state.isMobile;
     this.responsiveUIMultiple = this.getResponsiveUIMultiple(this.props.responsiveView);
     this.moreOptionsItems = null;
+    this.vr = false;
+    if (this.props.controller && this.props.controller.videoVrSource) {
+      this.vr = this.props.controller.videoVrSource;
```
Comment (jpfranco) [inline js/components/controlBar.js]: Is there a reason why we copy this to local state instead of using the prop directly? Any changes to this object will not be propagated to the component unless it is remounted.

File: tests/components/accessibilityControls-test.js [hunk 15]
```diff
@@ -117,8 +117,25 @@ describe('AccessibilityControls', function () {
     };
 
     var accessibilityControls = new AccessibilityControls(controllerMock);
-    accessibilityControls.handleKey(mockEvent);
-    accessibilityControls.handleKey(mockEvent);
+    accessibilityControls.keyEventDown(mockEvent);
+    accessibilityControls.keyEventDown(mockEvent);
+  });
+
+  it('tests "A" key', function () {
```
Comment (jpfranco) [inline tests/components/accessibilityControls-test.js]: Some of the older unit tests are not very good. I would recommend not basing new ones off of them. Latest changes from master should have a slightly better approach, please use this as a basis.

File: tests/components/accessibilityControls-test.js [hunk 16]
```diff
@@ -117,8 +117,25 @@ describe('AccessibilityControls', function () {
...
+      },
+      moveToDirection: function() {},
+    };
+
+    var mockEvent = {
+      keyCode: 65,
+    };
+
+    var accessibilityControls = new AccessibilityControls(controllerMock);
+    accessibilityControls.keyEventDown(mockEvent);
```
Comment (jpfranco) [inline tests/components/accessibilityControls-test.js]: Would recommend testing specific behavior rather than just executing the method.

File: tests/components/controlBar-test.js [hunk 17]
```diff
@@ -112,7 +112,136 @@ describe('ControlBar', function () {
     TestUtils.Simulate.click(fullscreenButton);
     expect(fullscreenToggled).toBe(true);
   });
+  
+  it('render one stereo button if content vr', function () {
+    var mockController = {
```
Comment (jpfranco) [inline tests/components/controlBar-test.js]: Please use `baseMockController` and `baseMockProps` properties instead of declaring new ones each time.

File: tests/components/controlBar-test.js [hunk 18]
```diff
@@ -112,7 +112,136 @@ describe('ControlBar', function () {
...
+                  playerState={CONSTANTS.STATE.PLAYING}
+                  isLiveStream={mockProps.isLiveStream} />
+    );
+    
+    var toggleStereoButton = TestUtils.findRenderedDOMComponentWithClass(DOM, 'oo-vr-stereo-button');
+    expect(typeof toggleStereoButton).toBe('object');
+  });
+  
+  it('not render stereo button if content not vr', function () {
+    var mockController = {
```
Comment (jpfranco) [inline tests/components/controlBar-test.js]: Same comment as above, we shouldn't re-declare these each time.

File: tests/components/controlBar-test.js [hunk 19]
```diff
@@ -112,7 +112,136 @@ describe('ControlBar', function () {
...
+                  componentWidth={500}
+                  playerState={CONSTANTS.STATE.PLAYING}
+                  isLiveStream={mockProps.isLiveStream} />
+    );
+    
+    var toggleStereoButtons = TestUtils.scryRenderedDOMComponentsWithClass(DOM, 'oo-vr-stereo-button');
+    expect(toggleStereoButtons.length).toBe(0);
+  });
+  
+  it('enter stereo mode', function () {
```
Comment (jpfranco) [inline tests/components/controlBar-test.js]: Please use naming convention for unit tests: `should perform x action` or `should do x thing`

**Comment by jpfranco** [comment]
@IvanSakharovskii for newer unit tests please try to use (or improve, if possible) the patterns in `volumeControls-test.js and `html5skin-test.js` or the bottom section (titled 'API') of `accessibilityControls-test.js`. Some of the older unit tests are not great, so moving forward we're trying to avoid repeating those old patterns.

File: js/components/accessibilityControls.js (hunk 1 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 2 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 3 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 4 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 5 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 6 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/accessibilityControls.js (hunk 8 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/components/directionControl.js (hunk 9 above)
Comment (KseniiaSokolovaEpam) [reply js/components/directionControl.js]: \+

File: js/views/playingScreen.js (hunk 11 above)
Comment (KseniiaSokolovaEpam) [reply js/views/playingScreen.js]: \+

File: js/components/controlBar.js (hunk 14 above)
Comment (KseniiaSokolovaEpam) [reply js/components/controlBar.js]: Thank you for the comment. We will change it in userstory 1424

Comment (KseniiaSokolovaEpam) [reply js/components/controlBar.js]: Thank you for the comment. We will change it in userstory 1424

File: tests/components/controlBar-test.js (hunk 17 above)
Comment (KseniiaSokolovaEpam) [reply tests/components/controlBar-test.js]: Now this part of the code is removed from the userstory. We add props to baseMockController in userstory 1424. Thank you.

File: js/components/directionControl.js (hunk 10 above)
Comment (KseniiaSokolovaEpam) [reply js/components/directionControl.js]: No, it should not be accessible via keyboard. We have special keyboard buttons for control.

File: js/views/playingScreen.js (hunk 12 above)
Comment (KseniiaSokolovaEpam) [reply js/views/playingScreen.js]: \+

File: scss/components/_vr-controls.scss (hunk 13 above)
Comment (KseniiaSokolovaEpam) [reply scss/components/_vr-controls.scss]: This part of the code is removed from this userstory and will be changed in newer userstory with using fonts instead of images.

File: tests/components/controlBar-test.js (hunk 19 above)
Comment (yvinokurov) [reply tests/components/controlBar-test.js]: this is fixed in feature / PLAYER-1424

File: tests/components/controlBar-test.js (hunk 18 above)
Comment (yvinokurov) [reply tests/components/controlBar-test.js]: this is fixed in userstory 1424

File: tests/components/accessibilityControls-test.js (hunk 16 above)
Comment (KseniiaSokolovaEpam) [reply tests/components/accessibilityControls-test.js]: ok, but it looks like it is enough in this situation.

File: tests/components/accessibilityControls-test.js (hunk 15 above)
Comment (KseniiaSokolovaEpam) [reply tests/components/accessibilityControls-test.js]: ok, we will, thanks.

File: js/components/controlBar.js [hunk 20]
```diff
@@ -310,7 +306,7 @@ var ControlBar = React.createClass({
       fullscreenIcon = "expand";
       fullscreenAriaLabel = CONSTANTS.ARIA_LABELS.FULLSCREEN;
     }
-
+
```
Comment (aldigun) [inline js/components/controlBar.js]: let's remove extra whitespaces on all the files/

File: js/components/controlBar.js [hunk 21]
```diff
@@ -98,7 +94,7 @@ var ControlBar = React.createClass({
     evt.preventDefault();
     this.props.controller.toggleFullscreen();
   },
-
+
```
Comment (aldigun) [inline js/components/controlBar.js]: whitespace

File: js/components/controlBar.js [hunk 22]
```diff
@@ -474,8 +470,8 @@ var ControlBar = React.createClass({
           <Tooltip enabled={isTooltipEnabled} text={Utils.getLocalizedString(this.props.language, CONSTANTS.SKIN_TEXT.SHARE, this.props.localizableStrings)} responsivenessMultiplier={this.responsiveUIMultiple} bottom={this.responsiveUIMultiple * this.props.skinConfig.controlBar.height} alignment={alignment} />
         </a>
       }).bind(this),
-
-      "fullscreen": (function (alignment) {
+
```
Comment (aldigun) [inline js/components/controlBar.js]: whitespace

File: js/components/controlBar.js [hunk 23]
```diff
@@ -617,6 +613,7 @@ var ControlBar = React.createClass({
     return returnStyles;
   },
 
+
```
Comment (aldigun) [inline js/components/controlBar.js]: whitespace

File: js/controller.js [hunk 24]
```diff
@@ -270,17 +271,15 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
...
+      this.accessibilityControls = this.accessibilityControls || new AccessibilityControls(this); //keyboard support
       this.state.screenToShow = CONSTANTS.SCREEN.INITIAL_SCREEN;
-
-      if (this.videoVr) {
-        this.mb.subscribe(OO.EVENTS.VR_DIRECTION_CHANGED, 'customerUi', _.bind(this.setViewingDirection, this));
-      }
     },
-
-    setVideoVr: function(event, obj) {
+
```
Comment (aldigun) [inline js/controller.js]: whitespace

File: js/views/playingScreen.js [hunk 25]
```diff
@@ -12,22 +12,23 @@ var React = require('react'),
     Watermark = require('../components/watermark'),
     ResizeMixin = require('../mixins/resizeMixin'),
     CONSTANTS = require('../constants/constants');
+    ViewControlsVr = require('../components/viewControlsVr'),
+    _ = require('underscore');
```
Comment (aldigun) [inline js/views/playingScreen.js]: underscore is not necessary in this file.

Comment (aldigun) [reply js/views/playingScreen.js]: There is _.isNumber() usage in this file..
Should we change this to use typeof X === "number", thenw e can remove _ ?

File: js/views/playingScreen.js [hunk 26]
```diff
@@ -203,17 +192,32 @@ var PlayingScreen = React.createClass({
...
+  /**
+   * @param paramName {string} "yaw", "pitch"
+   * @returns {number}
+   */
+  getViewingDirectionParamValue: function(paramName) {
+    var viewingDirectionYaw = 0;
+    if (this.props.controller &&
+      this.props.controller.state &&
+      this.props.controller.state.viewingDirection &&
+      _.isNumber(this.props.controller.state.viewingDirection[paramName])) {
```
Comment (aldigun) [inline js/views/playingScreen.js]: usage of underscore, should we change this to typeof X === "number" ?

Comment (aldigun) [reply js/views/playingScreen.js]: also, should we do a null check on this.props.controller.state.viewingDirection[paramName]?
Or typeof can do null check?

Comment (aldigun) [reply js/views/playingScreen.js]: Confirmed that typeof handles null check, so we are good.

File: js/views/playingScreen.js (hunk 25 above)
Comment (yvinokurov) [reply js/views/playingScreen.js]: fix

File: js/views/playingScreen.js (hunk 26 above)
Comment (yvinokurov) [reply js/views/playingScreen.js]: fix

File: js/components/controlBar.js (hunk 20 above)
Comment (yvinokurov) [reply js/components/controlBar.js]: fix

File: js/components/accessibilityControls.js [hunk 27]
```diff
@@ -3,33 +3,44 @@ var Utils = require('./utils');
...
+
+  document.addEventListener("keydown", this.keyEventDown);
+  document.addEventListener("keyup", this.keyEventUp);
 };
 
 AccessibilityControls.prototype = {
   cleanUp : function() {
-    document.removeEventListener('keydown', this.keyEvent);
+    document.removeEventListener("keydown", this.keyEventDown);
+    document.removeEventListener("keyup", this.keyEventUp);
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Not a big issue, but the rest of the file uses single quotes. We should try to keep coding style consistent.

File: js/components/accessibilityControls.js [hunk 28]
```diff
@@ -61,6 +72,53 @@ AccessibilityControls.prototype = {
...
 
+  keyEventUp: function(e) {
+    if (!this.controller.state.accessibilityControlsEnabled) { return; }
+    var targetTagName = this.getTargetTagName(e);
+    var charCode = e.which || e.keyCode;
+    this.moveVrToDirection(e, charCode, false, targetTagName);  //stop rotate 360
+  },
+
+  /**
+   *
```
Comment (jpfranco) [inline js/components/accessibilityControls.js]: Missing description

File: js/components/directionControlVr.js [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+var DirectionControlVr = React.createClass({
+  getInitialState: function() {
+    return {
+      isTouched: false,
+    };
+  },
+
+  handleEvent: function (ev) {
+    var rotate = ev.type == 'mousedown' || ev.type == 'touchstart';
```
Comment (jpfranco) [inline js/components/directionControlVr.js]: Can we use `===` here?

File: js/views/playingScreen.js [hunk 30]
```diff
@@ -64,18 +64,6 @@ var PlayingScreen = React.createClass({
     }
   },
 
-  handlePlayerMouseUp: function(event) {
```
Comment (jpfranco) [inline js/views/playingScreen.js]: Why was this removed?

File: js/components/accessibilityControls.js (hunk 27 above)
Comment (jpfranco) [reply js/components/accessibilityControls.js]: Actually never mind, it looks like they are all double quotes now.  We

File: js/views/playingScreen.js (hunk 30 above)
Comment (KseniiaSokolovaEpam) [reply js/views/playingScreen.js]: There were two the same functions.

File: js/components/directionControlVr.js (hunk 29 above)
Comment (KseniiaSokolovaEpam) [reply js/components/directionControlVr.js]: \+

File: js/components/accessibilityControls.js (hunk 28 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/controller.js (hunk 24 above)
Comment (KseniiaSokolovaEpam) [reply js/controller.js]: \+

File: js/components/accessibilityControls.js [hunk 31]
```diff
@@ -62,6 +73,59 @@ AccessibilityControls.prototype = {
...
+  getTargetTagName: function(e) {
+    var targetTagName = "";
+    if (e.target && typeof e.target.tagName === "string") {
+      targetTagName = e.target.tagName.toLowerCase();
+    }
+    return targetTagName;
+  },
+
+  /**
+   *
```
Comment (aeng7) [inline js/components/accessibilityControls.js]: Missing description.

File: js/components/accessibilityControls.js [hunk 32]
```diff
@@ -62,6 +73,59 @@ AccessibilityControls.prototype = {
...
+  keyEventUp: function(e) {
+    if (!this.controller.state.accessibilityControlsEnabled) {
+      return;
+    }
+    var targetTagName = this.getTargetTagName(e);
+    var charCode = e.which || e.keyCode;
+    this.moveVrToDirection(e, charCode, false, targetTagName);  //stop rotate 360
+  },
+
+  /**
```
Comment (aeng7) [inline js/components/accessibilityControls.js]: Missing description.

File: js/views/playingScreen.js [hunk 33]
```diff
@@ -203,17 +191,32 @@ var PlayingScreen = React.createClass({
...
     var degreesForPixelPitch = maxDegreesY / this.props.componentHeight;
-    var yaw = (this.props.controller.state.viewingDirection.yaw || 0) + dx * degreesForPixelYaw;
-    var pitch = (this.props.controller.state.viewingDirection.pitch || 0) + dy * degreesForPixelPitch;
+    var yaw = this.getViewingDirectionParamValue("yaw") + dx * degreesForPixelYaw;
+    var pitch = this.getViewingDirectionParamValue("pitch") + dy * degreesForPixelPitch;
     return [yaw, 0, pitch];
   },
 
+  /**
+   * @param paramName {string} "yaw", "pitch"
```
Comment (aeng7) [inline js/views/playingScreen.js]: Missing description.

File: js/components/accessibilityControls.js (hunk 32 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: The description is here already.

File: js/components/accessibilityControls.js (hunk 31 above)
Comment (KseniiaSokolovaEpam) [reply js/components/accessibilityControls.js]: \+

File: js/views/playingScreen.js (hunk 33 above)
Comment (KseniiaSokolovaEpam) [reply js/views/playingScreen.js]: \+

File: js/views/playingScreen.js (hunk 30 above)
Comment (jpfranco) [reply js/views/playingScreen.js]: Ok thanks

File: scss/mixins/_mixins.scss [hunk 34]
```diff
@@ -216,3 +216,24 @@
   top:0;
   left:0;
 }
+
+@function legacy-direction($direction) {
```
Comment (aeng7) [inline scss/mixins/_mixins.scss]: Why is this called legacy direction?

File: tests/components/accessibilityControls-test.js (hunk 15 above)
Comment (aeng7) [reply tests/components/accessibilityControls-test.js]: @jpfranco can you check if the new unit tests are satisfactory? Thanks!
